server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -235,10 +236,11 @@ private void getFromTranslog(
         final var retryingListener = listener.delegateResponse((l, e) -> {
             final var cause = ExceptionsHelper.unwrapCause(e);
             logger.debug("get_from_translog failed", cause);
+            // All of the following exceptions can be thrown if the shard is relocated
             if (cause instanceof ShardNotFoundException
                 || cause instanceof IndexNotFoundException
+                || cause instanceof IllegalIndexShardStateException
                 || cause instanceof AlreadyClosedException) {
```

Comment:
This AlreadyClosedException is thrown when the engine is closed due to relocation (not due to reset)